packages/service-library/src/servicelib/utils.py

```diff
@@ -175,3 +180,144 @@ def unused_port() -> int:
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
         s.bind(("127.0.0.1", 0))
         return cast(int, s.getsockname()[1])
+
+
+T = TypeVar("T")
+
+
+async def limited_as_completed(
+    awaitables: Iterable[Awaitable[T]] | AsyncIterable[Awaitable[T]],
+    *,
+    limit: int = _DEFAULT_LIMITED_CONCURRENCY,
+    tasks_group_prefix: str | None = None,
+) -> AsyncGenerator[asyncio.Future[T], None]:
+    """Runs awaitables using limited concurrent tasks and returns
+    result futures unordered.
+
+    Arguments:
+        awaitables -- The awaitables to limit the concurrency of.
+
+    Keyword Arguments:
+        limit -- The maximum number of awaitables to run concurrently.
+                0 or negative values disables the limit. (default: {1})
+        tasks_group_prefix -- The prefix to use for the name of the asyncio tasks group.
+                             If None, no name is used. (default: {None})
+
+    Returns:
+        nothing
+
+    Yields:
+        Future[T]: the future of the awaitables as they appear.
+
+
+    """
+    try:
+        awaitable_iterator = aiter(awaitables)  # type: ignore[arg-type]
+        is_async = True
+    except TypeError:
+        assert isinstance(awaitables, Iterable)  # nosec
+        awaitable_iterator = iter(awaitables)  # type: ignore[assignment]
+        is_async = False
+
+    completed_all_awaitables = False
+    pending_futures: set[asyncio.Future] = set()
+
+    try:
+        while pending_futures or not completed_all_awaitables:
+            while (
+                limit < 1 or len(pending_futures) < limit
+            ) and not completed_all_awaitables:
+                try:
+                    aw = (
+                        await anext(awaitable_iterator)
```

Comment:
MINOR: instead of check everytime `is_async`, you can prepare all variants in common variables and use them later e.g.

```python
try
    awaitable_iterator = aiter(awaitables)
    next_func = anext
except:
    awaitable_iterator = iter(awaiteables)
    next_func = next

```